cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1862,3 +1888,428 @@ private predicate flowsTo(
 predicate flowsTo(Node source, Node sink, Configuration configuration) {
   flowsTo(_, _, source, sink, configuration)
 }
+
+private module FlowExploration {
+  private predicate callableStep(DataFlowCallable c1, DataFlowCallable c2, Configuration config) {
+    exists(Node node1, Node node2 |
+      jumpStep(node1, node2, config)
+      or
+      additionalJumpStep(node1, node2, config)
+      or
+      // flow into callable
+      viableParamArg(_, node2, node1)
+      or
+      // flow out of an argument
+      exists(ParameterNode p |
+        parameterValueFlowsToUpdate(p, node1) and
+        viableParamArg(_, p, node2.(PostUpdateNode).getPreUpdateNode())
+      )
+      or
+      // flow out of a callable
+      exists(DataFlowCall call, ReturnKind kind |
+        getReturnPosition(node1) = viableReturnPos(call, kind) and
+        node2 = getAnOutNode(call, kind)
+      )
+    |
+      c1 = node1.getEnclosingCallable() and
+      c2 = node2.getEnclosingCallable() and
+      c1 != c2
+    )
+  }
+
+  private predicate interestingCallableSrc(DataFlowCallable c, Configuration config) {
+    exists(Node n | config.isSource(n) and c = n.getEnclosingCallable())
+    or
+    exists(DataFlowCallable mid |
+      interestingCallableSrc(mid, config) and callableStep(mid, c, config)
+    )
+  }
+
+  private newtype TCallableExt =
+    TCallable(DataFlowCallable c, Configuration config) { interestingCallableSrc(c, config) } or
+    TCallableSrc()
```

Comment:
I'm just looking for the shortest distance to _some_ source, so this just unifies all the source nodes.  There's no real gain in repeating the distance calculation for each source node, and the additional computation time might be non-trivial.

Reply:
But couldn't you just define
```
private predicate callableExtSrc(TCallableExt ce) {
  exists(Configuration config, Node n |
    ce = TCallable(n.getEnclosingCallable(), config) |
    config.isSource(node)
  )
}
```
?

Reply:
That would cause the BFS inside the HOP to be executed a number of times equal to the number of callables with sources instead of just once.  Note that the result of the HOP gives the distance parameterized by the source.

Reply:
OK, I didn't know that would result in multiple BFSs.